### ISSUE TITLE
ProGuard/R8 Problem

### ISSUE THREAD

**Issue by apm-mike-klimek** [trbnb/MvvmBase#15]
Hi,

I have an error at runtime, when running ProGuard or R8 on a project that uses MvvmBase 1.4.1:
`java.lang.UnsupportedOperationException: Type not found: class de.trbnb.mvvmbase.BaseViewModel$lifecycle$1`.
This happens when a viewmodel is created at `de.trbnb.mvvmbase.BaseViewModel.<init>(BaseViewModel.kt:45)`.

These are my ProGuard rules:
```proguard
-keepattributes Signature, InnerClasses, EnclosingMethod
-keep public class de.trbnb.mvvmbase.** { *; }
# Prevent a BuildLoader error. Not needed when running with R8.
-keep interface kotlin.reflect.jvm.internal.impl.builtins.BuiltInsLoader
-keep class kotlin.reflect.jvm.internal.impl.serialization.deserialization.builtins.BuiltInsLoaderImpl
```

I think this is caused by the following bug: https://discuss.kotlinlang.org/t/proguard-cant-process-this-simple-code/7054
There is an open issue at the Kotlin bug tracker: https://youtrack.jetbrains.com/issue/KT-16084

Maybe this can be solved in the meantime, by refactoring the code around `BaseViewModel.lifecycle` a bit.

**Comment by trbnb** [comment]
Yeah, it seems to be the Lifecycle in the `BaseViewModel`. I think it's better in its own file anyway, so I'll move that.

After that you'll probably need more Proguard rules for the reflection based features to work. I think I'll add them to the README.

**Comment by trbnb** [comment]
I have released a version 1.4.2 that should fix this issue.
